### ISSUE TITLE
Tree Dump Files

### ISSUE THREAD

**Issue by therealcain** [compiler-explorer/compiler-explorer#2547]
I wanted to see new flags in GCC to understand this compiler a bit more in-depth, and i encountered a flag that i think it could be a great addition to the compiler explorer website.

Assembly code is hard to understand to many people, and if we would like to know if something was optimized we must look at the assembly code and understand it, So apparently there's a flag for both GCC and Clang ( I do not know about MSVC or any other compiler ), to see what the compiler did in a more human-readable code, rather than assembly.

Here is an example in GCC: ( compiled like `gcc test.c -O3 -fdump-tree-all` )
Say we have a recursive function and we would like to see if it was optimized into a loop ( this is very common among tail-recursion, but as far as i know there are cases that the compiler does not always optimize it, so we must look at the assembly if it actually did an optimization ):
```c
// test.c
//
#include <stdio.h>

void infinite_loop()
{
    printf("Hello World!\n");
    infinite_loop();
}

int main() { infinite_loop(); }
```

Obviously in this case we can run the program and wait to see if a stack overflow was occurred, if it's not the compiler must've optimized it to a loop, also the assembly code for this specific function is very easy to understand, it's just a simple example to emphasis what i mean.
After compiling this code with the flag `-fdump-tree-all` many dump files were created ( if you want you can read about them here: https://gcc.gnu.org/onlinedocs/gcc/Developer-Options.html ), but the one that we desire is the *optimized* file, it contains the same code after optimization, which looks like this:
```c
;; Function infinite_loop (infinite_loop, funcdef_no=23, decl_uid=2510, cgraph_uid=24, symbol_order=23) (executed once)

Merging blocks 3 and 4
infinite_loop ()
{
  <bb 2> [local count: 107374]:

  <bb 3> [local count: 1073741824]:
  __builtin_puts (&"Hello World!"[0]);
  goto <bb 3>; [100.00%]

}
```

We can tell that our `infinite_loop` recursion call was optimized to `goto`, and we can already see that our `printf("Hello World\n")` was optimized to a faster better suited print call which is `puts`.
This is very much easier to understand compared to assembly code, especially when dealing with big weird functions.
Here is the assembly code for comparison:
```asm
.LC0:
        .string "Hello World!"
loop():
        sub     rsp, 8
.L2:
        mov     edi, OFFSET FLAT:.LC0
        call    puts
        jmp     .L2
main:
        sub     rsp, 8
        call    loop()
```

We can also take this a step further and show a nice diagram of what's happening, by compiling with `gcc test.c -O3 -fdump-tree-all-graph` ( Since it created our optimization file as dot file, we need to convert it to an image: `dot -Tpng optimized.dot > optimized.png` ):
![image](https://user-images.githubusercontent.com/22757058/113173490-17405e00-9252-11eb-965f-f877892d826c.png)

For Clang there are also options for dumping files like that, i did not try it, but I'm aware that it's possible with `-Xclang`.

There are many cases that we could use it either in C or C++, and i think it could be a very handy and cool feature to be added to compiler explorer.

**Comment by mattgodbolt** [comment]
Discussed a little on the Discord. Some current work:

* The "GCC Tree/RTL Viewer", available from "Add new..." on the compiler view.
* The CFG viewer available from "Add new..." and then picking "Graph Output"
* For clang there's also the AST Output and IR output available on the same menu,

The dot output from GCC is separate and new. @cppchedy suggests we might be able to convert dot output to vis.js and show it. I think I'll close this bug in favour of a "support dot output from GCC" bug.

**Comment by mattgodbolt** [comment]
Closed in favour of #2552
